### PR TITLE
Composer map to follow a visibility preset

### PR DESCRIPTION
**Pull request by wonder-sk**
(fixes Redmine [#13418](http://hub.qgis.org/issues/13418))

This adds a new option in composer map properties:
"Follow visibility preset" with a combo box to choose the active preset.

This is an alternative to "lock layers" (and "lock layer styles") functionality
which would just copy preset's configuration, while the new option links to preset.

The difference is that when a preset is updated, composer map will automatically
pick the new configuration when following the preset, while there is no update
if "lock layers" (and "lock layer styles") option is used.

### CODE REVIEW

**Comment by nirvn** [comment]
@wonder-sk , ahh, I always thought presets linked to map items should have worked like this. Thanks for getting us there :smile:

**Comment by NathanW2** [comment]
+1 good idea

**Comment by nyalldawson** [comment]
@wonder-sk 

The actually following works for me, but the combobox is not updated when new presets are added. I think this is because nothing ever emits the QgsVisibilityPresetCollection::presetsChanged() signal...

**Comment by nyalldawson** [comment]
@wonder-sk  should be fixed by b51d70ab956e9c67ab30ac9d746ab2d8de064978

However, the current preset is lost when presets are changed - eg remove a preset, and the current composer map preset will be reset to none even if a different preset was removed.
It also happens when a preset changes - since that's also now emitting presetsChanged..

**Comment by DelazJ** [comment]
@wonder-sk a big thank you for tackling this, one of the most hated issue in my daily work for recent releases :)
Looking forward to testing it!

**Comment by wonder-sk** [comment]
Thanks a lot @nyalldawson for testing and the presetsChanged() fix!

**Comment by nyalldawson** [comment]
@wonder-sk I've been thinking - should we remove the presetsChanged signal from preset updates? Ie only emit it when presets are added/deleted/renamed? Otherwise this signal will be continually emitted whenever the styles changes. 

I'm thinking yes (and update the docs to reflect this), and then add a different signal which is emitted just when a preset is updated.

**Comment by wonder-sk** [comment]
@nyalldawson yeah that makes sense to me. Maybe even worth renaming the existing signal to something more obvious, e.g. presetsListChanged? (vs e.g. presetUpdated(QString))

**Comment by 3nids** [comment]
@wonder-sk great work, seems to work as expected!

Just not 100% sure about the UI.

I think it would make things clearer in a new dedicated collapsible group box.
I would also rephrase a bit.
What do you think of:

* Layers
  * [ ] Follow visibility preset 
  * [ ] Lock layers
     * [ ] Lock styles for layers

**Comment by DelazJ** [comment]
Once again, thanks @wonder-sk for fixing this. I'll try it extensively later in the day (first try lets me think that it seems to follow preset as expected) but just few comments below about UI.
I agree with @3nids about the ideas of:
* grouping these options in a frame
* an offset of the "Lock styles for layers" option, given that it's only available when the layers are locked. User will easily understand what it adds to the previous option.
* rephrasing (map item is already implicit when selecting the map)

Another thing about the UI is the use of a simple combobox to list presets (to follow) and a iconic combobox to also list preset (for the block layers option). For consistency, better use same widget imo.

If needed, I will comment how it works later (here, in #13418 or open new issue report if not really related to what was discussed in #13418)

File: python/core/composer/qgscomposermap.sip [hunk 1]
```diff
@@ -223,6 +223,27 @@ class QgsComposerMap : QgsComposerItem
     /** Stores the current layer styles into style overrides. @note added in 2.8 */
     void storeCurrentLayerStyles();
 
+    /** Whether the map should follow a visibility preset. If true, the layers and layer styles
+     * will be used from given preset name (configured with setFollowVisibilityPresetName() method).
+     * This means when preset's settings are changed, the new settings are automatically
+     * picked up next time when rendering, without having to explicitly update them.
+     * At most one of the flags keepLayerSet() and followVisibilityPreset() should be enabled
+     * at any time since they are alternative approaches - if both are enabled,
```
Comment (DelazJ) [inline python/core/composer/qgscomposermap.sip]: I may misunderstand something but how could both be enabled, given that they are mutually exclusive?

Comment (wonder-sk) [inline python/core/composer/qgscomposermap.sip]: They are mutually exclusive in GUI, however not in API (that would need a backwards incompatible change). So the note is just to make people aware. In 3.0 we could switch to an enumeration of possible modes.

**Pull request merged**